### DISCUSSION TITLE
Coupling OpenFAST with CFD code written in FORTRAN 90

### DISCUSSION THREAD

**Discussion by brunolop** [OpenFAST/openfast#1253] [Q&A]
Hello,

I’m a PhD student trying to develop a two-way (weak) coupling between OpenFAST and our in-house CFD code, for onshore applications. Both codes are written in FORTRAN 90 language. 

The CFD would be used to model the wind flow around the rotor using Actuator Line Model and interpolate the wind speed at blade nodes, which will be transferred to OpenFAST. The aeroelastic code would compute the aerodynamical loads, structural deformation and control the turbine, and finally the aerodynamical loads and blade nodes positions would be transferred back to the CFD.

In order to accomplish the coupling, I’m considering the following options:

1) Create an interface in the Fortran CFD code to communicate with C++ API (which will eventually call FORTRAN subroutines of OpenFAST) and follow an approach similar to SOWFA. 
2) Develop a “FORTRAN API” for OpenFAST and avoid interfaces between FORTRAN and C++.
3) Integrate the CFD code with OpenFAST following an approach similar to FAST-Farm.

What do you think about these ideas? Would you recommend any of them, or a different one? Any suggestions would be much appreciated. 

Best regards,
Bruno

**Comment by AgreeU** [comment]
Hello,
I'm working on creating the interface between OpenFAST and in-house code with C++ API. But I met a problem on compiling the code with cmake. Have you ever met this before?

Regards,
Yang
![image](https://github.com/OpenFAST/openfast/assets/118701870/39a1062d-9897-4486-ac1e-63ac083cd9ca)
